core/src/main/java/hudson/Launcher.java

```diff
@@ -393,24 +385,23 @@ public Proc start() throws IOException {
 
         /**
          * Starts the process and waits for its completion.
-         * This method stores internal context and hence it is synchronized
          * @return Return code of the invoked process
          * @throws IOException Operation error (e.g. remote call failure)
          * @throws InterruptedException The process has been interrupted
          */
-        public synchronized int join() throws IOException, InterruptedException {
+        public int join() throws IOException, InterruptedException {
             // The logging around procHolderForJoin prevents the preliminary object deallocation we saw in JENKINS-23271
-            try {
-                procHolderForJoin = start();
-                LOGGER.log(Level.FINER, "Started the process {0}", procHolderForJoin);
-                final int returnCode = procHolderForJoin.join();
-                if (LOGGER.isLoggable(Level.FINER)) {
-                    LOGGER.log(Level.FINER, "Process {0} has finished with the return code {1}", new Object[] {procHolderForJoin, returnCode});
-                }
-                return returnCode;
-            } finally {
-                procHolderForJoin = null;
+            final Proc procHolderForJoin = start();
+            LOGGER.log(Level.FINER, "Started the process {0}", procHolderForJoin);
+            final int returnCode = procHolderForJoin.join();
+            if (LOGGER.isLoggable(Level.FINER)) {
+                LOGGER.log(Level.FINER, "Process {0} has finished with the return code {1}", new Object[] {procHolderForJoin, returnCode});
+            }
+            if (procHolderForJoin.isAlive()) { // Should never happen but this forces Proc to not be removed and early GC by escape analysis
+                assert false : "Process not finished after call to join() completed";
```

Comment:
So the contract of finally means that the JVM cannot evaluate it out of order, and then by not bothering with an assert, rather running to warning log level then escape analysis cannot remove the if... plus while it should never happen, we want to know if it does... an assertion error is not a good way to have that happen...